qiime2/core/type/signature.py

```diff
@@ -779,3 +790,34 @@ def _make_hashable(self, collection):
             return collection
 
         return tuple(new_collection)
+
+
+class CaptureHolder:
+    CAPTURE_HOLDER_DEFAULT = None
+
+    def __init__(self, name, value, type, provenance):
+        self._set = False
+        self._name = name
+        self._value = value
+        self._type = type
+        self._provenance = provenance
+
+    @property
+    def is_set(self):
+        return self._set or self._value != CaptureHolder.CAPTURE_HOLDER_DEFAULT
+
+    @property
+    def value(self):
+        return self._value
+
+    def set_value(self, value):
+        if self._set:
+            raise ValueError(f'Value already set to {self._value}')
+
+        if value in self._type:
```

Comment:
`None` is a special case here, it isn't in any primitive type, so we shuffle it along on a special track in signature. It is however a valid value here because the parameter is optional by definition (and always `None` as default).

```suggestion
        if value is None or value in self._type:
```